### DISCUSSION TITLE
Change the osu! UI audio options on Linux

### DISCUSSION THREAD

**Discussion by airi019** [ppy/osu#28782] [Ideas] [answered]
**CLARIFICATION: THIS IS A UI CHANGE PROPOSAL. I'M NOT TELLING YOU TO CHANGE THE SOUND LIBRARY!!!!!**

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

On Linux, it is pretty much a custom that userland apps shouldn't be able to choose their own output devices. Instead that is done through system apps.

**Rationale:**

![Screenshot from 2024-07-08 22-56-25](https://github.com/ppy/osu/assets/155326638/7552d55e-2f61-41bd-a51f-ea6f017fb49a)

**This essentially makes all of these audio options pointless!**

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

**My proposal is to remove this "output device" option and default to Pipewire**.

![Screenshot from 2024-07-08 17-56-18](https://github.com/ppy/osu/assets/155326638/66c0d58f-8d77-4d16-80cd-52e0babedc1e)

**Nearly all Linux distros now use Pipewire as a audio backend including Linux Mint starting on version 22 which is on beta and is going to release soon.** 

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

**NOT A PROBLEM:**

1 - Outputing to the audio device directly is not possible since osu doesn't have permission to do so, **although a compatibility layer is going to take care of that**(UNLESS you have multiple audio cards and decide to dedicate one to osu!).

2 - The sound server runs on the cpu but added latency is negligible since you can adjust audio offset.

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

**What can you do with Pipewire:**

1 - You can send any Linux app audio to these Android apps through your local network:

https://github.com/roc-streaming/roc-droid
https://github.com/kaytat/SimpleProtocolPlayer
https://github.com/mkckr0/audio-share

The Android device then functions as a wireless speaker or headphones if you plug one. **This adds latency but you can still adjust audio offset.**

2 - It allows creating and removing connections between applications and/or devices to reroute flow of audio, video and MIDI data to where it is needed.

https://flathub.org/apps/org.pipewire.Helvum
https://flathub.org/apps/org.rncbc.qpwgraph

**Comment by airi019** [comment]
![Screenshot from 2024-07-08 18-33-33](https://github.com/ppy/osu/assets/155326638/3701c86b-dede-466f-a33c-d361a9348d44)
device busy error example

**Comment by sineplusx** [comment]


**Reply by airi019** [reply]
1 - https://flathub.org/apps/io.github.yairm210.unciv, https://flathub.org/apps/io.bassi.Amberol

2 -  who? even debian is pipewire only now

3 - so what? get used to it?

**Reply by Stoppedpuma** [reply]
> also, choosing alsa devices ("direct access to audio device") will give "device busy" errors since osu is not allowed to do that.

@airisama019 I don't mean to be a dick but if you don't understand how linux audio servers work then you probably shouldn't be telling developers how to implement them. Not to mention, just because a few programs do something doesn't mean it's necessarily correct, let alone should be followed by every other program.

> 2 - who? even debian is pipewire only now

One of the main points of running an operating system like Linux is the ability of choice? Some people may prefer one thing over another or are running older systems such as peppy running Debian 10 on his own test vm. There's also people in the process of trying to run osu! under BSDs with tools such as Linuxulator so who knows what other doors are open? Removing this option is only bad in my eyes.

**Reply by airi019** [reply]
dude youre just making it harder for new users; I thought osu was about simplicity.

**Reply by Stoppedpuma** [reply]
Because choosing an audio output is hard? If you can't figure out how to choose an output then Linux probably isn't for you in the first place.

**Reply by airi019** [reply]
thats why everyone is trying to fix it

**Reply by Stoppedpuma** [reply]
I would like to see this "everyone" as I haven't seen this complaint in the osu! repo once before this.

**Reply by airi019** [reply]
not osu but linux instead

**Reply by airi019** [reply]
and most osu players don't even know you can complain here

**Reply by Stoppedpuma** [reply]
> and most osu players don't even know you can complain here

Most players who don't know you can complain here probably aren't using Linux either.

> not osu but linux instead

Still applies, I've never seen a request as ridiculous as removing basic audio functionality.

**Reply by airi019** [reply]
basic audio functionality that you cant use
congrats

**Reply by Stoppedpuma** [reply]
I know right, it would be awful if changing my audio device to one which pipewire wasn't using didn't work right? Wait...

**Reply by airi019** [reply]
what are u talking about. Pipewire is always using all audio devices

**Reply by airi019** [reply]
and like i said, there are apps to redirect any app output to whatever audio device you want.

**Reply by Feodor0090** [reply]
> basic audio functionality that you cant use

your system is misconfigured. I route sound from osu! via "raw" alsa device to remove sound server from audio output "chain", because it does nothing but introducing latency, and that works.

**Reply by airi019** [reply]
> > basic audio functionality that you cant use
> 
> your system is misconfigured. I route sound from osu! via "raw" alsa device to remove sound server from audio output "chain", because it does nothing but introducing latency, and that works.

and what about other apps? just mute them?

**Reply by Feodor0090** [reply]
> there are apps to redirect any app output

Any software that works with audio introduces latency. If you want ~0 latency (for osu!, you always want), your choice is ASIO or ALSA.

**Reply by airi019** [reply]
> > there are apps to redirect any app output
> 
> Any software that works with audio introduces latency. If you want ~0 latency (for osu!, you always want), your choice is ASIO or ALSA.

idk what u talking about dude i play on default debian gnome wayland pipewire and i can get osumania rainbows with offset -50

**Reply by Feodor0090** [reply]
> and what about other apps? just mute them?

You don't need sound from other apps when you play <ins>rhythm</ins> game.

**Reply by airi019** [reply]
> > and what about other apps? just mute them?
> 
> You don't need sound from other apps when you play rhythm game.

omg dude

**Reply by Stoppedpuma** [reply]
> what are u talking about. Pipewire is always using all audio devices

Odd that doing the exact thing I mentioned works then? Maybe you don't know what you're talking about?

[1720480572.webm](https://github.com/ppy/osu/assets/58333920/2248e3a1-6fa8-44b6-a58d-b8911361055a)

Since this is obviously bait I'm going to quit wasting my time, in the near impossible odds that you are being genuine then you should follow my recommendation above and learn what you're talking about before trying to tell devs how to implement it.

**Reply by airi019** [reply]
> > what are u talking about. Pipewire is always using all audio devices
> 
> Odd that doing the exact thing I mentioned works then? Maybe you don't know what you're talking about?
> 
>  1720480572.webm 
> Since this is obviously bait I'm going to quit wasting my time, in the near impossible odds that you are being genuine then you should follow my recommendation above and learn what you're talking about before trying to tell devs how to implement it.

![image](https://github.com/ppy/osu/assets/155326638/45cb85dd-bace-4951-b26c-4f320354aacf)
like I said you can set output device whenever you want

and you should switch to linux before supporting the platform

**Reply by airi019** [reply]
> > also, choosing alsa devices ("direct access to audio device") will give "device busy" errors since osu is not allowed to do that.
> 
> @airisama019 I don't mean to be a dick but if you don't understand how linux audio servers work then you probably shouldn't be telling developers how to implement them. Not to mention, just because a few programs do something doesn't mean it's necessarily correct, let alone should be followed by every other program.
> 
> > 2 - who? even debian is pipewire only now
> 
> One of the main points of running an operating system like Linux is the ability of choice? Some people may prefer one thing over another or are running older systems such as peppy running Debian 10 on his own test vm. There's also people in the process of trying to run osu! under BSDs with tools such as Linuxulator so who knows what other doors are open? Removing this option is only bad in my eyes.



> > and most osu players don't even know you can complain here
> 
> Most players who don't know you can complain here probably aren't using Linux either.
> 
> > not osu but linux instead
> 
> Still applies, I've never seen a request as ridiculous as removing basic audio functionality.



> I know right, it would be awful if changing my audio device to one which pipewire wasn't using didn't work right? Wait...



> > what are u talking about. Pipewire is always using all audio devices
> 
> Odd that doing the exact thing I mentioned works then? Maybe you don't know what you're talking about?
> 
>  1720480572.webm 
> Since this is obviously bait I'm going to quit wasting my time, in the near impossible odds that you are being genuine then you should follow my recommendation above and learn what you're talking about before trying to tell devs how to implement it.

stop making it harder for regular players like me to complain about the game; we shouldn't need a degree for it and all I wanted was to make it easier for new linux players to play the game.

**Comment by Stoppedpuma** [comment]
osu! currently uses BASS for audio, BASS doesn't have native support for Pipewire and Pulseaudio servers but instead uses ALSA. You can find a PR for supporting SDL3 as an audio backend [here](https://github.com/ppy/osu-framework/pull/6002).

**Reply by airi019** [reply]
is there a working build yet?

**Reply by Stoppedpuma** [reply]
Unless you compile yourself, no.

**Reply by bdach** [reply]
As usual I will restate that the chance of that PR making it in is... slim. BASS works out of the box on all other platforms. Supporting thirty different audio backends is a linux-unique situation and as such I'm personally okay (as a linux user) to disregard it until stuff breaks.

**Reply by airi019** [reply, marked as answer]
> As usual I will restate that the chance of that PR making it in is... slim. BASS works out of the box on all other platforms. Supporting thirty different audio backends is a linux-unique situation and as such I'm personally okay (as a linux user) to disregard it until stuff breaks.

OK I understand, so basically I need a crippled, dedicated version of linux for osu. Got it. But at least hide these options if the detected platform is Linux. Thank you.
